src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -26,26 +26,75 @@
 package jdk.incubator.foreign;
 
 import jdk.internal.foreign.NativeMemorySegmentImpl;
+import jdk.internal.foreign.PlatformLayouts;
 import jdk.internal.foreign.Utils;
 import jdk.internal.foreign.abi.SharedUtils;
 
-import java.nio.ByteOrder;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
 import java.nio.charset.Charset;
 import java.util.Objects;
 import java.util.function.Consumer;
 
+import static jdk.internal.foreign.PlatformLayouts.*;
+
 /**
- * A set of utilities for working with libraries using the C language/ABI
+ * A foreign linker specializing for C Application Binary Interface (ABI) calling conventions.
+ * Instances of this interface can be used to link foreign functions in native libraries that
+ * follow the JVM's target platform C ABI.
+ *
+ * <p>There are two components that go into linking a foreign function: a method type, and
+ * a function descriptor. The method type, consisting of a set of <em>carrier</em> types constitute the
+ * set of carrier types, which, together, specify the Java signature which clients must adhere to when
+ * calling the underlying foreign function. The function descriptor contains a set of memory layouts which,
+ * together, specify the foreign function signature and classification information (via custom layout attributes),
+ * so that linking can take place. Memory layout attributes are used in the function descriptor
+ * to attach ABI classification meta-data to memory layouts, which are required for linking.
+ * Clients of this API should build function descriptors using the predefined memory layout constants
+ * (based on a subset of the built-in types provided by the C language), found in this interface;
+ * a failure to do so might result in linkage errors, given that linking requires additional classification
+ * information to determine, for instance, how arguments should be loaded into registers during a
+ * foreign function call.</p>
```

Comment:
I think this paragraph is general enough that it would work very well on the top ForeignLinker interface (replace C with any other language). I also wonder if we should add an example here - e.g. for `strlen` ?